File: src/main/java/dev/sky_lock/pocketlifevehicle/vehicle/model/ModelBuilder.java
package dev.sky_lock.pocketlifevehicle.vehicle.model;

import java.util.List;

/**
 * @author sky_lock
 */

public class ModelBuilder {
    private final String id;
    private String name;
    private List<String> lores;
    private float maxFuel;
    private MaxSpeed maxSpeed;
    private ModelItem item;
    private Capacity capacity;
    private CollideBox collideBox;
    private float height;
    private SteeringLevel steeringLevel;
    private Sound sound;
    private ModelPosition position;

    public ModelBuilder(String id) {
        this.id = id;
    }

    public static ModelBuilder of(String id) {
        return new ModelBuilder(id);
    }

    public ModelBuilder name(String name) {
        this.name = name;
        return this;
    }

    public ModelBuilder item(ModelItem item) {
        this.item = item;
        return this;
    }

    public ModelBuilder lores(List<String> lores) {
        this.lores = lores;
        return this;
    }

    public ModelBuilder maxFuel(float maxFuel) {
        this.maxFuel = maxFuel;
        return this;
    }

    public ModelBuilder maxSpeed(MaxSpeed maxSpeed) {
        this.maxSpeed = maxSpeed;
        return this;
    }

    public ModelBuilder capacity(Capacity capacity) {
        this.capacity = capacity;
        return this;
    }

    public ModelBuilder collideBox(float baseSide, float height) {
        this.collideBox = new CollideBox(baseSide, height);
        return this;
    }

    public ModelBuilder height(float height) {
        this.height = height;
        return this;
    }

    public ModelBuilder steering(SteeringLevel steeringLevel) {
        this.steeringLevel = steeringLevel;
        return this;
    }

    public ModelBuilder sound(Sound sound) {
        this.sound = sound;
        return this;
    }

    public ModelBuilder modelPosition(ModelPosition position) {
        this.position = position;
        return this;
    }

    public Model build() {
        if (name == null || maxSpeed == null || item == null ||
                capacity == null || collideBox == null || steeringLevel == null ||
                sound == null || position == null) {
            throw new NullPointerException();
        }
        return new Model(id, item, name, lores, maxFuel, maxSpeed, capacity, steeringLevel, collideBox, height, sound, position);
    }

}

File: src/main/kotlin/dev/sky_lock/pocketlifevehicle/vehicle/model/ModelBuilder.kt
package dev.sky_lock.pocketlifevehicle.vehicle.model

/**
 * @author sky_lock
 */
class ModelBuilder(private val id: String) {
    private var name: String? = null
    private var lore: List<String>? = null
    private var spec: Spec? = null
    private var itemOption: ItemOption? = null
    private var isBig = false
    private var collideBox: CollideBox? = null
    private var height = 0f
    private var sound: Sound? = null
    fun name(name: String): ModelBuilder {
        this.name = name
        return this
    }

    fun lore(lore: List<String>): ModelBuilder {
        this.lore = lore
        return this
    }

    fun spec(spec: Spec): ModelBuilder {
        this.spec = spec
        return this
    }

    fun item(itemOption: ItemOption): ModelBuilder {
        this.itemOption = itemOption
        return this
    }

    fun big(isBig: Boolean): ModelBuilder {
        this.isBig = isBig
        return this
    }

    fun collideBox(baseSide: Float, height: Float): ModelBuilder {
        collideBox = CollideBox(baseSide, height)
        return this
    }

    fun height(height: Float): ModelBuilder {
        this.height = height
        return this
    }

    fun sound(sound: Sound): ModelBuilder {
        this.sound = sound
        return this
    }

    fun build(): Model {
        if (name == null || spec == null || itemOption == null || collideBox == null || sound == null) {
            throw IllegalStateException()
        }
        return Model(id, name!!, lore!!, spec!!, itemOption!!, collideBox!!, isBig, height, sound!!)
    }

    companion object {
        fun of(id: String): ModelBuilder {
            return ModelBuilder(id)
        }
    }

}